shiny/render/_dataframe.py

```diff
@@ -258,29 +402,334 @@ class data_frame(Renderer[DataFrameResult]):
+                data = self.data().copy(deep=False)
+                for cell_patch in self.cell_patches():
+                    data.iat[  # pyright: ignore[reportUnknownMemberType]
+                        cell_patch.row_index,
+                        cell_patch.column_index,
+                    ] = cell_patch.value
+
+                return data
+
+        self.data_patched = self_data_patched
+
+    def _get_session(self) -> Session:
+        if self._session is None:
+            raise RuntimeError(
+                "The data frame's session can only be accessed within a reactive context"
+            )
+        return self._session
+
+    def set_cell_update_fn(self, fn: CellUpdateFn) -> Self:
+        self._cell_update_fn = fn
+        return self
+
+    def set_cells_update_fn(self, fn: CellsUpdateFn) -> Self:
+        self._cells_update_fn = fn
+        return self
+
+    def _init_handlers(self) -> None:
+        async def _set_cell_update_default(
+            *,
+            info: CellUpdateInfo,
+            # row_index: int,
+            # column_index: int,
+            # value: str,
+            # prev: str,
+            **kwargs: Any,
+        ) -> str:
+            return info["value"]
+
+        async def _set_cells_update_default(
+            *,
+            infos: list[CellUpdateInfo],
+            **kwargs: Any,
+        ):
+            with reactive.isolate():
+                formatted_values: list[Any] = []
+                for update_info in infos:
+                    # row_index = update_info["row_index"]
+                    # column_index = update_info["column_index"]
+                    # value = update_info["value"]
+                    # prev = update_info["prev"]
+
+                    formatted_value = await self._cell_update_fn(info=update_info)
+                    #     row_index=row_index,
+                    #     column_index=column_index,
+                    #     value=value,
+                    #     prev=prev,
+                    # )
+                    # TODO-barret; check type here?
+                    # TODO-barret; The return value should be coerced by pandas to the correct type
+                    formatted_values.append(formatted_value)
+
+                return formatted_values
+
+        self.set_cell_update_fn(_set_cell_update_default)
+        self.set_cells_update_fn(_set_cells_update_default)
+        # self._add_message_handlers()
+
+    # To be called by session's output_handler message handler on this data_frame instance
+    @output_dispatch_handler
+    # Do not change this method name unless you update corresponding code in `/js/dataframe/`!!
+    async def _handle_cells_update(self, update_infos: list[CellUpdateInfo]):
+
+        # Make new array to trigger reactive update
+        patches = [p for p in self.cell_patches()]
+
+        # Call user's cell update method to retrieve formatted values
+        updated_raw_values = await self._cells_update_fn(infos=update_infos)
```

Comment:
`_cells_update_fn` should return a list of objects with row, col, value. And don't check that the input and output lengths are the same.

Update the JS to match.